tiledb/sm/array_schema/domain.h

```diff
@@ -192,9 +199,14 @@ class Domain {
    *
    * @param buff The buffer to deserialize from.
    * @param version The array schema version.
-   * @return Status
+   * @return Status and Domain
    */
-  Status deserialize(ConstBuffer* buff, uint32_t version);
+  static std::tuple<Status, optional<std::shared_ptr<Domain>>> deserialize(
+      ConstBuffer* buff,
+      uint32_t version,
+      uint64_t cell_num_per_tile = 0,
+      Layout cell_order = Layout::ROW_MAJOR,
+      Layout tile_order = Layout::ROW_MAJOR);
```

Comment:
tl;dr version: (1) This function has an extra argument (2) This function should not have default arguments. (3) It needs expanded documentation.

Digging into this code, this class has a mismatch between its class design and its storage format. The storage format does not contain cell and tile order, but the class does. Similarly, the member variable `cell_num_per_tile_` is not present on disk but is always computed from other values.

`cell_num_per_tile` should not be an argument here. Its corresponding member variable `cell_num_per_tile_` should always be computed in the constructor. It's initialized to zero in the (legacy) default constructor and then actually initialized in `init()`. There's nothing in the existing code that supplies a value to this member in any way that doesn't originate in the function that computes it.

Cell and tile order are different. They are in storage in the array schema and then duplicated here. I don't understand whatever reason this is the case, or even if it has a good reason, but for the scope of this PR we're going to deal with it. These arguments must be provided by the array schema. In the case of unit tests, provide your own explicit value to substitute for the ones that would otherwise be coming from an array schema. There should not be default values on this function for the convenience of the unit tests. The unit tests need to follow the needs of the code, not lead them.

Lastly, we need documentation for why these "extra" arguments are need by deserialization to construct a fully-formed object. It's not at all obvious. It's arguably a defect in the storage format, but we're not changing that in the current PR. Documentation is the extent to which we can deal with this issue at present.